[proposal]: Add sensitivity check to prefetcher

devisscher opened the pull request:
## Description

Proposing to add a sensitivity check to Prefetcher. This allows us to trigger prefetching with a little more accuracy by ensuring the mouse over is intentional. We can use the change in mouse position to calculate whether or not the user intends to navigate to said url. 

Heavily inspired by: https://github.com/natelindev/react-use-hoverintent/blob/master/src/index.ts

### Demo/Explanation

https://user-images.githubusercontent.com/9765013/220665662-5ae03a51-8cba-4fd0-a6e1-1f60e5c3dc4e.mp4

### 🎩 'ing instructions

[Spin URL](https://admin.web.constellation-6t51.thomas-devisscher.us.spin.dev/store/shop1)

1. Go to admin
2. Hover over sidebar navigation menu items
3. You shouldn't see network requests being fired if you hover over items quickly
4. You should see network requests being fired if you stop on an item for more than 2 seconds


<details><summary>

#### Rolling your own instance ⬇️

</summary>


##### How to work with Quilt packages

Some steps to work with Quilt packages or any other package locally/in your spin instance.

1. Create a constellation which includes the repo Quilt

```ymal
schema: v1
extend: shopify
repos:
  - web
  - argus-server
  - quilt
```

```
spin up constellation.yml
```

2. Open the quilt repo in spin and make your changes to the package
3. Publish the package locally with publish:local script. this puts the package in the yalc store

```
PKG_NAME=react-async yarn publish:local
```

4. Install yalc globally: yarn global add yalc, if yalc command is not found, export the path to global node_modules.

```
export PATH="$(yarn global bin):$PATH"
```

5. `yalc add @shopify/react-async && restart`  will grab the package from the yalc store and add it to your web repo then restart web.

</details>

<details><summary>

#### Slowing things down ⬇️

</summary>

1. Go to `packages/react-async/src/Prefetcher.tsx`
2. Increase the `INTENTION_DELAY_MS` to 2000
3. On line 122 add a console.log('prefetching')
4. On line 123 add a console.log('not prefetching')

Then publish the package, add it into web (Follow rolling your own instance instructions), and navigate to the admin. When hovering over an item, and pausing on it, you should see prefetching in the console after 2 seconds. If you hover over quickly, prefetching shouldn't fire.

</details>

I am available to pair or demo this at anytime! 🍐

### Before this can be implemented
- [x] Test mousemove event and ensure prefetch is triggered/not triggered when threshold is reached/not reached
- [x] Ensure all cases work as expected in web
- [x] Remove console logs/ cruft
- [ ] Discuss configuration, do we need to expose the sensitivity prop to consumers of this package?

devisscher commented:
/snapit

knames reviewed (approved):
nice work!

patrickDouglas commented:
@devisscher Is there any way to reload your spin instance?  I couldn't find it in the Spin Control dashboard, or I would have done it myself.

atesgoral commented on packages/react-async/src/Prefetcher.tsx:
```diff
@@ -92,6 +109,22 @@ class ConnectedPrefetcher extends React.PureComponent<Props, State> {
+    }
+    // Calculate the change of the mouse position
+    // If it is smaller than the sensitivity, we can assume that the user is intending on visiting the link
+    if (Math.abs(this.fX - iX) + Math.abs(this.fY - iY) < SENSITIVITY) {
```
But then you may want to change the default to 15 pixels or something.

atesgoral replied:
(This will give you the actual distance travelled, when x and y are changing at the same time)

devisscher commented:
Thank you for the reviews! @atesgoral I've updated the code to address your comments, thank you for the suggestions! 🙇‍♂️ I also updated the tests, and the condition for calling the compare function. (Instead of checking for MouseEvent instance, we can check for the existence of clientX and clientY.
